implement nationalities system / added countries.json

SNAILX808 opened the pull request:
### Description

implemented nationalities system

---

### Implementation Details

- Integrated nationality data into player data structure in es_extended
- Nationality stored as string identifier in database for flexibility
- stored nationalities in countries.json
- Added server-side security check to validate only countries listed in countries.json


---

### Usage Example

```lua
-- client-side usage 
ESX.GetPlayerData().nationality

-- server-side usage
-- get nationality
xPlayer.getNationality()
or 
xPlayer.nationality
-- set nationality
xPlayer.setNationality("esx")
```

---

### PR Checklist

-   [x] My commit messages and PR title follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard.
-   [x] My changes have been tested locally and function as expected.
-   [x] My PR does not introduce any breaking changes.
-   [x] I have provided a clear explanation of what my PR does, including the reasoning behind the changes and any relevant context.

SNAILX808 committed:
implement nationalities system / added countries.json

SNAILX808 committed:
updated

- converted countries to lua table
- applied numeric loop in isValidNationality function
- added countries.lua

Zuntie commented on [core]/esx_identity/client/main.lua:
```diff
@@ -27,6 +29,7 @@ end)
 
 RegisterNUICallback("ready", function(_, cb)
     ready = true
+    SendNUIMessage({ type = "config", data = Config })
```
> Use the `cb` param to send the data. So something like `cb({ data = Config })` No need to have another event sending the config, unnecessary.

Yep, that's a good idea. Thank you very much! Slipped my mind.

Zykem replied:
Use the cb param to send the data. So something like cb({ data = Config }) No need to have another event sending the config, unnecessary.

Zykem commented on [core]/esx_identity/server/main.lua:
```diff
@@ -114,6 +116,23 @@ local function checkHeightFormat(height)
+local function isValidNationality(nationality)
+    for i = 1, #Config.countryList do
+        local country = Config.countryList[i]
+        if string.lower(country.value) == string.lower(nationality) then
```
Make `nationality` lowercase on top of the function. If you keep it like that, it will convert the same string to lowercase x amount of times before it finds the country (useless overhead). So first define the a variable like `normalizedNationality` and compare it with `string.lower(country.value)`

Zykem commented on [core]/esx_identity/server/main.lua:
```diff
@@ -114,6 +116,23 @@ local function checkHeightFormat(height)
     return numHeight >= Config.MinHeight and numHeight <= Config.MaxHeight
 end
 
+local function checkNationalityFormat(nationality)
```
Add proper type annotations here.

Zykem replied:
The same here, please documentate the code. Add a proper return annotation, the name `checkNationalityFormat` does not tell you what it returns.

Zykem commented on [core]/esx_identity/server/main.lua:
```diff
@@ -114,6 +116,23 @@ local function checkHeightFormat(height)
+    return false
+end
+
+local function isValidNationality(nationality)
```
Add proper type annotations here.

Zykem replied:
Add `---@return boolean IsValidNationality` (or similar) annotation, the user checking this code needs to know whats going on by reading the annotations.

Zykem commented on [core]/esx_identity/countries.lua:
```diff
@@ -0,0 +1,196 @@
+return {
```
Please add `Palestine` and `Kosovo` to the countries list. Flags: 🇵🇸 🇽🇰 , Labels: `Palestine`, `Kosovo`

Zykem commented on [core]/esx_identity/esx_identity.sql:
```diff
@@ -3,5 +3,6 @@ ALTER TABLE `users`
 	ADD COLUMN `sex` VARCHAR(1) NULL DEFAULT NULL,
-	ADD COLUMN `height` INT NULL DEFAULT NULL
+	ADD COLUMN `height` INT NULL DEFAULT NULL,
+	ADD COLUMN `nationality` VARCHAR(30) NULL DEFAULT NULL
```
Increase the maximum amount of characters for `nationality` to 52. 30 is too short even for some countries in that list.

SNAILX808 committed:
optimization

SNAILX808 committed:
added 2 missed countries / increase amount of characters for nationality column

Zuntie commented on [core]/esx_identity/client/main.lua:
```diff
@@ -2,8 +2,7 @@ local loadingScreenFinished = false
 local timecycleModifier = "hud_def_blur"
-local countryList = json.decode(Config.countryList)
- --TODO: make the ui fetch countries from countryList
+ --TODO: make the ui fetch countries from countryList using Config.countryList
```
Remove this comment - the whole Config will be sent to the UI :) .

Kenshiin13 committed:
Merge pull request #1748 from esx-framework/set-metadata

feat(es_extended): add xPlayer.updatePlayerData

SNAILX808 committed:
Merge pull request #1 from esx-framework/dev

Dev

SNAILX808 committed:
implemented 2 new functions in (es_extended/server/classes/player.lua)

- setNationality func to set the nationality and used updatePlayerData to sync the nationality on the client-side
- getNationality func to return nationality as string value

- removed the parts that use state bag

Zykem committed:
feat(migrations) Add nationality migration

SNAILX808 committed:
refactor(esx_identity) removed ESX.require

SNAILX808 committed:
Merge branch 'dev' of https://github.com/SNAILX808/esx_core into dev

Zykem committed:
refactor: Directly send back config instead of using SendNUIMessage

Zykem reviewed (approved):
Tested, LGTM

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/esx-framework/esx_core?pullRequest=1746) <br/>All committers have signed the CLA.
